### PR TITLE
new model

### PR DESCRIPTION
**Pull request by dgornjakovic**
A pull request to depict the differences between the Thales model and the original Synthetix model.
To sum up the differences:

- Bidding phase is removed
- For every sUSD deposited into a binary market the user gets 1 long option and 1 short option
- Options can be traded as soon as the market is created up until maturity
- Anyone can mint more options by depositing the adequate amount of sUSD
- Refunds are cancelations are removed as those were only possible during bidding phase

### CODE REVIEW

File: contracts/BinaryOption.sol [hunk 1]
```diff
@@ -1,16 +1,15 @@
 pragma solidity ^0.5.16;
 
 // Inheritance
-import "./interfaces/IERC20.sol";
+import "synthetix-2.43.1/contracts/interfaces/IERC20.sol";
```
Comment (dgornjakovic) [inline contracts/BinaryOption.sol:4]: all synthetix contracts are imported in this manner

File: contracts/BinaryOption.sol [hunk 2]
```diff
@@ -19,125 +18,50 @@ contract BinaryOption is IERC20, IBinaryOption {
 
     /* ========== STATE VARIABLES ========== */
 
-    string public constant name = "SNX Binary Option";
-    string public constant symbol = "sOPT";
+    string public name;
+    string public symbol;
```
Comment (dgornjakovic) [inline contracts/BinaryOption.sol:22]: Name and Symbol will be created dynamically now as it was not user friendly to have all Options be named "sOPT"

File: contracts/BinaryOption.sol [hunk 3]
```diff
@@ -19,125 +18,50 @@ contract BinaryOption is IERC20, IBinaryOption {
...
     uint8 public constant decimals = 18;
 
     BinaryOptionMarket public market;
 
-    mapping(address => uint) public bidOf;
-    uint public totalBids;
-
```
Comment (dgornjakovic) [inline contracts/BinaryOption.sol:30]: There is no longer a concept of bidding

File: contracts/BinaryOption.sol [hunk 4]
```diff
@@ -19,125 +18,50 @@ contract BinaryOption is IERC20, IBinaryOption {
...
     mapping(address => uint) public balanceOf;
     uint public totalSupply;
 
     // The argument order is allowance[owner][spender]
     mapping(address => mapping(address => uint)) public allowance;
 
-    // Enforce a 1 cent minimum bid balance
-    uint internal constant _MINIMUM_BID = 1e16;
+    // Enforce a 1 cent minimum amount
+    uint internal constant _MINIMUM_AMOUNT = 1e16;
```
Comment (dgornjakovic) [inline contracts/BinaryOption.sol:34]: Variable is still used for minimal amount that is mintable.

File: contracts/BinaryOption.sol [hunk 5]
```diff
@@ -19,125 +18,50 @@ contract BinaryOption is IERC20, IBinaryOption {
...
-    function _totalClaimableSupply(uint exercisableDeposits) internal view returns (uint) {
-        uint _totalSupply = totalSupply;
-        // We'll avoid throwing an exception here to avoid breaking any dapps, but this case
-        // should never occur given the minimum bid size.
-        if (exercisableDeposits <= _totalSupply) {
-            return 0;
-        }
-        return exercisableDeposits.sub(_totalSupply);
-    }
+    bool public initialized = false;
```
Comment (dgornjakovic) [inline contracts/BinaryOption.sol:38]: Minimal proxy pattern

File: contracts/BinaryOption.sol [hunk 6]
```diff
@@ -19,125 +18,50 @@ contract BinaryOption is IERC20, IBinaryOption {
...
-        return _totalClaimableSupply(exercisableDeposits);
+    function initialize(
+        string memory _name,
+        string memory _symbol
+    ) public {
+        require(!initialized, "Binary Option Market already initialized");
+        initialized = true;
+        name = _name;
+        symbol = _symbol;
+        market = BinaryOptionMarket(msg.sender);
```
Comment (dgornjakovic) [inline contracts/BinaryOption.sol:48]: Minimal proxy pattern

File: contracts/BinaryOption.sol [hunk 7]
```diff
@@ -19,125 +18,50 @@ contract BinaryOption is IERC20, IBinaryOption {
...
-        bidOf[bidder] = _requireMinimumBid(bidOf[bidder].add(newBid));
-        totalBids = totalBids.add(newBid);
-    }
-
-    // This must only be invoked during bidding.
-    function refund(address bidder, uint newRefund) external onlyMarket {
-        // The safe subtraction will catch refunds that are too large.
-        bidOf[bidder] = _requireMinimumBid(bidOf[bidder].sub(newRefund));
-        totalBids = totalBids.sub(newRefund);
-    }
```
Comment (dgornjakovic) [inline contracts/BinaryOption.sol:116]: Refunds are  no longer applicable

File: contracts/BinaryOption.sol [hunk 8]
```diff
@@ -19,125 +18,50 @@ contract BinaryOption is IERC20, IBinaryOption {
...
-        emit Transfer(address(0), claimant, claimable);
-        emit Issued(claimant, claimable);
+    function mint(address minter, uint amount) external onlyMarket {
+        _requireMinimumAmount(amount);
+        totalSupply = totalSupply.add(amount);
+        balanceOf[minter] = balanceOf[minter].add(amount); // Increment rather than assigning since a transfer may have occurred.
 
-        return claimable;
+        emit Transfer(address(0), minter, amount);
+        emit Issued(minter, amount);
```
Comment (dgornjakovic) [inline contracts/BinaryOption.sol:64]: There is no longer a concept of claiming. For every sUSD that is deposited the user gets 1 long option and 1 short option (minus the 1% minting fee)

File: contracts/BinaryOptionMarket.sol [hunk 9]
```diff
@@ -29,13 +28,7 @@ contract BinaryOptionMarket is Owned, MixinResolver, IBinaryOptionMarket {
         BinaryOption short;
     }
 
-    struct Prices {
-        uint long;
-        uint short;
-    }
-
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarket.sol:36]: Prices are no longer a concept. Trading is done P2P via 0x orderbooks.

File: contracts/BinaryOptionMarket.sol [hunk 10]
```diff
@@ -49,99 +42,77 @@ contract BinaryOptionMarket is Owned, MixinResolver, IBinaryOptionMarket {
     /* ========== STATE VARIABLES ========== */
 
     Options public options;
-    Prices public prices;
     Times public times;
     OracleDetails public oracleDetails;
     BinaryOptionMarketManager.Fees public fees;
-    BinaryOptionMarketManager.CreatorLimits public creatorLimits;
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarket.sol:56]: There is no longer a concept of an skew, so minimal skew is gone as well. Minimal requirement of 1000 sUSD to create a market is kept to ensure some initial orderbook liquidity.

File: contracts/BinaryOptionMarket.sol [hunk 11]
```diff
@@ -49,99 +42,77 @@ contract BinaryOptionMarket is Owned, MixinResolver, IBinaryOptionMarket {
...
+    IAddressResolver public resolver;
 
-    // `deposited` tracks the sum of open bids on short and long, plus withheld refund fees.
+    // `deposited` tracks the sum of all deposits minus the withheld fees.
     // This must explicitly be kept, in case tokens are transferred to the contract directly.
     uint public deposited;
+    uint public initialMint;
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarket.sol:53]: Creator will not pay the fee for initial mint, so we keep the variable for later.

File: contracts/BinaryOptionMarket.sol [hunk 12]
```diff
@@ -49,99 +42,77 @@ contract BinaryOptionMarket is Owned, MixinResolver, IBinaryOptionMarket {
...
     address public creator;
     bool public resolved;
-    bool public refundsEnabled;
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarket.sol:63]: refunds are no longer a concept

File: contracts/BinaryOptionMarket.sol [hunk 13]
```diff
@@ -49,99 +42,77 @@ contract BinaryOptionMarket is Owned, MixinResolver, IBinaryOptionMarket {
...
 
     uint internal _feeMultiplier;
 
     /* ---------- Address Resolver Configuration ---------- */
 
-    bytes32 internal constant CONTRACT_SYSTEMSTATUS = "SystemStatus";
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarket.sol:69]: Thales system will not depend on Synthetix system

File: contracts/BinaryOptionMarket.sol [hunk 14]
```diff
@@ -49,99 +42,77 @@ contract BinaryOptionMarket is Owned, MixinResolver, IBinaryOptionMarket {
...
     bytes32 internal constant CONTRACT_EXRATES = "ExchangeRates";
     bytes32 internal constant CONTRACT_SYNTHSUSD = "SynthsUSD";
-    bytes32 internal constant CONTRACT_FEEPOOL = "FeePool";
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarket.sol:72]: Fee pool is now a configurable address

File: contracts/BinaryOptionMarket.sol [hunk 15]
```diff
@@ -49,99 +42,77 @@ contract BinaryOptionMarket is Owned, MixinResolver, IBinaryOptionMarket {
...
+        _mint(creator, initialMint);
 
-    function resolverAddressesRequired() public view returns (bytes32[] memory addresses) {
-        addresses = new bytes32[](4);
-        addresses[0] = CONTRACT_SYSTEMSTATUS;
-        addresses[1] = CONTRACT_EXRATES;
-        addresses[2] = CONTRACT_SYNTHSUSD;
-        addresses[3] = CONTRACT_FEEPOOL;
+        // Note: the ERC20 base contract does not have a constructor, so we do not have to worry
+        // about initializing its state separately
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarket.sol:105]: Minimal proxy pattern to initialize

File: contracts/BinaryOptionMarket.sol [hunk 16]
```diff
@@ -378,16 +213,6 @@ contract BinaryOptionMarket is Owned, MixinResolver, IBinaryOptionMarket {
...
-    function _checkCreatorLimits(uint longBid, uint shortBid) internal view {
-        uint totalBid = longBid.add(shortBid);
-        require(creatorLimits.capitalRequirement <= totalBid, "Insufficient capital");
-        uint skewLimit = creatorLimits.skewLimit;
-        require(
-            skewLimit <= longBid.divideDecimal(totalBid) && skewLimit <= shortBid.divideDecimal(totalBid),
-            "Bids too skewed"
-        );
-    }
-
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarket.sol:390]: this is moved to the manager

File: contracts/BinaryOptionMarket.sol [hunk 17]
```diff
@@ -404,72 +229,38 @@ contract BinaryOptionMarket is Owned, MixinResolver, IBinaryOptionMarket {
...
+        uint valueAfterFees = value.multiplyDecimalRound(_feeMultiplier);
 
-        uint _deposited = _incrementDeposited(value);
-        _sUSD().transferFrom(msg.sender, address(this), value);
+        _mint(msg.sender, valueAfterFees);
 
-        (uint longTotalBids, uint shortTotalBids) = _totalBids();
-        _updatePrices(longTotalBids, shortTotalBids, _deposited);
+        _incrementDeposited(value);
+        _manager().transferSusdTo(msg.sender, address(this), value);
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarket.sol:250]: all bidding and price logic is obsolete. The mint method takes the sUSD amount and mints 1 long and 1 short option minus the 1% fee for every sUSD.

File: contracts/BinaryOptionMarket.sol [hunk 18]
```diff
@@ -485,60 +276,25 @@ contract BinaryOptionMarket is Owned, MixinResolver, IBinaryOptionMarket {
         // in the contract will be sufficient to cover these transfers.
         IERC20 sUSD = _sUSD();
 
-        uint _deposited = deposited;
-        uint poolFees = _deposited.multiplyDecimalRound(fees.poolFee);
-        uint creatorFees = _deposited.multiplyDecimalRound(fees.creatorFee);
+        uint _depositedSubMint = deposited.sub(initialMint);
+        uint poolFees = _depositedSubMint.multiplyDecimalRound(fees.poolFee);
+        uint creatorFees = _depositedSubMint.multiplyDecimalRound(fees.creatorFee);
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarket.sol:281]: The initial mint is not subject to fees

File: contracts/BinaryOptionMarket.sol [hunk 19]
```diff
@@ -485,60 +276,25 @@ contract BinaryOptionMarket is Owned, MixinResolver, IBinaryOptionMarket {
...
         _decrementDeposited(creatorFees.add(poolFees));
-        sUSD.transfer(_feePool().FEE_ADDRESS(), poolFees);
+        sUSD.transfer(_manager().feeAddress(), poolFees);
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarket.sol:283]: store the fee address in the manager

File: contracts/BinaryOptionMarket.sol [hunk 20]
```diff
@@ -485,60 +276,25 @@ contract BinaryOptionMarket is Owned, MixinResolver, IBinaryOptionMarket {
...
-        emit OptionsClaimed(msg.sender, longOptions, shortOptions);
-        return (longOptions, shortOptions);
-    }
-
-    function claimOptions() external returns (uint longClaimed, uint shortClaimed) {
-        return _claimOptions();
-    }
-
-    function exerciseOptions() external returns (uint) {
+    function exerciseOptions() external afterMaturity returns (uint) {
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarket.sol:291]: only exercise is applicable for winning options

File: contracts/BinaryOptionMarket.sol [hunk 21]
```diff
@@ -583,28 +339,15 @@ contract BinaryOptionMarket is Owned, MixinResolver, IBinaryOptionMarket {
...
     }
 
-    function cancel(address payable beneficiary) external onlyOwner duringBidding {
-        (uint longTotalBids, uint shortTotalBids) = _totalBids();
-        (uint creatorLongBids, uint creatorShortBids) = _bidsOf(creator);
-        bool cancellable = longTotalBids == creatorLongBids && shortTotalBids == creatorShortBids;
-        require(cancellable, "Not cancellable");
-        _selfDestruct(beneficiary);
-    }
-
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarket.sol:593]: cancel concept is obsolete

File: contracts/BinaryOptionMarketFactory.sol [hunk 22]
```diff
@@ -1,64 +1,64 @@
...
+            strikePrice,
+            times,
+            initialMint,
+            fees
+        );
+        return bom;
+    }
+
+    /* ========== SETTERS ========== */
+    function setBinaryOptionMarketManager(address _binaryOptionMarketManager) public onlyOwner {
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarketFactory.sol:53]: Minimal proxy mastercopies addresses are stored in factory.

File: contracts/BinaryOptionMarketManager.sol [hunk 23]
```diff
@@ -39,16 +36,13 @@ contract BinaryOptionMarketManager is Owned, Pausable, MixinResolver, IBinaryOpt
...
-    }
-
     /* ========== STATE VARIABLES ========== */
 
+    address public feeAddress;
+
     Fees public fees;
     Durations public durations;
-    CreatorLimits public creatorLimits;
+    uint public capitalRequirement;
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarketManager.sol:45]: The only limit is now the 1000 sUSD to create a market

File: contracts/BinaryOptionMarketManager.sol [hunk 24]
```diff
@@ -58,66 +52,62 @@ contract BinaryOptionMarketManager is Owned, Pausable, MixinResolver, IBinaryOpt
 
     BinaryOptionMarketManager internal _migratingManager;
 
+    IAddressResolver public resolver;
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarketManager.sol:55]: Mixing resolver transformed into IAddressResolver

File: contracts/BinaryOptionMarketManager.sol [hunk 25]
```diff
@@ -58,66 +52,62 @@ contract BinaryOptionMarketManager is Owned, Pausable, MixinResolver, IBinaryOpt
...
+    }
 
-    function resolverAddressesRequired() public view returns (bytes32[] memory addresses) {
-        addresses = new bytes32[](4);
-        addresses[0] = CONTRACT_SYSTEMSTATUS;
-        addresses[1] = CONTRACT_SYNTHSUSD;
-        addresses[2] = CONTRACT_EXRATES;
-        addresses[3] = CONTRACT_BINARYOPTIONMARKETFACTORY;
+    function setFeeAddress(address _feeAddress) public onlyOwner {
+        feeAddress = _feeAddress;
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarketManager.sol:98]: fee address can be changed by governance when Thales matures

File: contracts/BinaryOptionMarketManager.sol [hunk 26]
```diff
@@ -236,70 +212,62 @@ contract BinaryOptionMarketManager is Owned, Pausable, MixinResolver, IBinaryOpt
...
+        uint maturity,
+        uint initialMint // initial sUSD to mint options for
     )
         external
         notPaused
         returns (
             IBinaryOptionMarket // no support for returning BinaryOptionMarket polymorphically given the interface
         )
     {
-        _systemStatus().requireSystemActive();
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarketManager.sol:249]: marketCreationEnabled and paused flags are enough to control downtimes if ever needed.

File: contracts/BinaryOptionMarketManager.sol [hunk 27]
```diff
@@ -236,70 +212,62 @@ contract BinaryOptionMarketManager is Owned, Pausable, MixinResolver, IBinaryOpt
...
-        _sUSD().transferFrom(msg.sender, address(market), initialDeposit);
+        totalDeposited = totalDeposited.add(initialMint);
+        _sUSD().transferFrom(msg.sender, address(market), initialMint);
 
-        emit MarketCreated(address(market), msg.sender, oracleKey, strikePrice, biddingEnd, maturity, expiry);
+        emit MarketCreated(address(market), msg.sender, oracleKey, strikePrice, maturity, expiry);
         return market;
     }
 
+    function transferSusdTo(address sender, address receiver, uint amount) external {
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarketManager.sol:258]: all sUSD transfers are done via manager to reduce the number of approvals.

File: contracts/BinaryOptionMarketManager.sol [hunk 28]
```diff
@@ -313,28 +281,6 @@ contract BinaryOptionMarketManager is Owned, Pausable, MixinResolver, IBinaryOpt
         }
     }
 
-    /* ---------- Upgrade and Administration ---------- */
-
-    function rebuildMarketCaches(BinaryOptionMarket[] calldata marketsToSync) external {
```
Comment (dgornjakovic) [inline contracts/BinaryOptionMarketManager.sol:318]: obsolete

File: contracts/OwnedWithInit.sol [hunk 29]
```diff
@@ -0,0 +1,38 @@
+pragma solidity ^0.5.16;
+
+contract OwnedWithInit {
```
Comment (dgornjakovic) [inline contracts/OwnedWithInit.sol:3]: a version of the Owner interface for MinimalProxy initialization.

File: contracts/BinaryOption.sol (hunk 6 above)
Comment (dgornjakovic) [reply contracts/BinaryOption.sol:48]: All the logic related to claiming options is gone, as options are minted into users wallet on deposit

File: contracts/BinaryOption.sol [hunk 30]
```diff
@@ -19,125 +18,50 @@ contract BinaryOption is IERC20, IBinaryOption {
...
+        market = BinaryOptionMarket(msg.sender);
     }
 
     /* ========== MUTATIVE FUNCTIONS ========== */
 
-    function _requireMinimumBid(uint bid) internal pure returns (uint) {
-        require(bid >= _MINIMUM_BID || bid == 0, "Balance < $0.01");
-        return bid;
+    function _requireMinimumAmount(uint amount) internal pure returns (uint) {
+        require(amount >= _MINIMUM_AMOUNT || amount == 0, "Balance < $0.01");
```
Comment (syncikin) [inline contracts/BinaryOption.sol:54]: why can you mint 0?

Comment (dgornjakovic) [reply contracts/BinaryOption.sol:54]: I will check, but as you can see the original version had bid 0 in it too.

Comment (dgornjakovic) [reply contracts/BinaryOption.sol:54]: I see there is even a unit tests in the original version that covered this and made sure bid zero is idempotent.

Comment (zyzek) [reply contracts/BinaryOption.sol:54]: The refund function also used to use this function. It allowed a user to refund their entire bid, but not refund an amount that would leave their bid at less than the required minimum.

**Pull request merged**